openapi3/internalize_refs.go

```diff
@@ -0,0 +1,389 @@
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+func parametersMapNames(s ParametersMap) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+func isExternalRef(ref string) bool {
+	return ref != "" && !strings.HasPrefix(ref, "#/components/")
+}
+
+func (doc *T) addSchemaToSpec(s *SchemaRef, refNameResolver RefNameResolver) string {
+	if s == nil || !isExternalRef(s.Ref) {
+		return ""
+	}
+
+	name := refNameResolver(s.Ref)
+
+	// I just added a more comprehensive check for this being a schema but people should really
+	// make sure their resolver returns unique names
+	val := 1
+	for {
+		// basic check, may need something more exhaustive...
+		if existing, ok := doc.Components.Schemas[name]; ok {
+			if existing.Value.Description != s.Value.Description || s.Value.Type != existing.Value.Type {
+				name = refNameResolver(s.Ref) + strconv.Itoa(val)
+				val++
+				continue
+			}
+			newRef := "#/components/schemas/" + name
+			if newRef == s.Ref {
+				s.Ref = ""
+			} else {
+				s.Ref = newRef
+			}
+			return s.Ref
+		}
+		break
+	}
+	if doc.Components.Schemas == nil {
+		doc.Components.Schemas = make(Schemas)
+	}
+	doc.Components.Schemas[name] = s.Value.NewRef()
+	s.Ref = "#/components/schemas/" + name
+	return s.Ref
+}
+
+func (doc *T) addParameterToSpec(p *ParameterRef, refNameResolver RefNameResolver) string {
+	if p == nil || !isExternalRef(p.Ref) {
+		return ""
+	}
+	name := refNameResolver(p.Ref)
+	if _, ok := doc.Components.Parameters[name]; ok {
+		p.Ref = "#/components/parameters/" + name
+		return p.Ref
+	}
+
+	if doc.Components.Parameters == nil {
+		doc.Components.Parameters = make(ParametersMap)
+	}
+	doc.Components.Parameters[name] = &ParameterRef{Value: p.Value}
+	p.Ref = "#/components/parameters/" + name
+	return p.Ref
+}
+
+func (doc *T) addHeaderToSpec(h *HeaderRef, refNameResolver RefNameResolver) string {
+	if h == nil || !isExternalRef(h.Ref) {
+		return ""
+	}
+	name := refNameResolver(h.Ref)
+	if _, ok := doc.Components.Headers[name]; ok {
+		h.Ref = "#/components/headers/" + name
+		return h.Ref
+	}
+	if doc.Components.Headers == nil {
+		doc.Components.Headers = make(Headers)
+	}
+	doc.Components.Headers[name] = &HeaderRef{Value: h.Value}
+	h.Ref = "#/components/headers/" + name
+	return h.Ref
+}
+
+func (doc *T) addRequestBodyToSpec(r *RequestBodyRef, refNameResolver RefNameResolver) string {
+	if r == nil || !isExternalRef(r.Ref) {
+		return ""
+	}
+	name := refNameResolver(r.Ref)
+	if _, ok := doc.Components.RequestBodies[name]; ok {
+		r.Ref = "#/components/requestBodies/" + name
+		return r.Ref
+	}
+	if doc.Components.RequestBodies == nil {
+		doc.Components.RequestBodies = make(RequestBodies)
+	}
+	doc.Components.RequestBodies[name] = &RequestBodyRef{Value: r.Value}
+	r.Ref = "#/components/requestBodies/" + name
+	return r.Ref
+}
+
+func (doc *T) addResponseToSpec(r *ResponseRef, refNameResolver RefNameResolver) string {
```

Comment:
```suggestion
func (doc *T) addResponseToSpec(r *ResponseRef, refNameResolver RefNameResolver) {
```
No call to this function actually use the returned value so let's drop that.